lib/libzfs/libzfs_dataset.c

```diff
@@ -486,18 +486,28 @@ make_dataset_handle(libzfs_handle_t *hdl, const char *path)
 
 	zhp->zfs_hdl = hdl;
 	(void) strlcpy(zhp->zfs_name, path, sizeof (zhp->zfs_name));
-	if (zcmd_alloc_dst_nvlist(hdl, &zc, 0) != 0) {
-		free(zhp);
-		return (NULL);
-	}
-	if (get_stats_ioctl(zhp, &zc) == -1) {
-		zcmd_free_nvlists(&zc);
-		free(zhp);
-		return (NULL);
-	}
-	if (make_dataset_handle_common(zhp, &zc) == -1) {
-		free(zhp);
-		zhp = NULL;
+
+	if (!hdl->libzfs_force_export) {
+		if (zcmd_alloc_dst_nvlist(hdl, &zc, 0) != 0) {
+			free(zhp);
+			return (NULL);
+		}
+		if (get_stats_ioctl(zhp, &zc) == -1) {
+			zcmd_free_nvlists(&zc);
+			free(zhp);
+			return (NULL);
+		}
+		if (make_dataset_handle_common(zhp, &zc) == -1) {
+			free(zhp);
+			zhp = NULL;
+		}
+	} else {
+		/*
+		 * Called from zpool_disable_datasets(), which uses mount
+		 * entries, so de facto the dataset is a ZFS.
```

Comment:
Not while `hdl->libzfs_force_export` is true.

Reply:
Then perhaps let's add that information, and the answer to my other question, to the comment to head off any confusion.

```
		 * Called from zpool_disable_datasets() which sets force
		 * export and uses mount entries, so de facto the dataset
		 * is a ZFS filesystem.  Furthermore, we need to avoid calling
		 * get_stats_ioctl() here since it result in zfs_ioc_objset_stats()->
		 * dmu_objset_hold() being called by the kernel which can
		 * potentially cause IO to be issued depending on what's
		 * currently cached in ARC.
```